xmtp_mls/src/groups/device_sync.rs

```diff
@@ -399,540 +358,514 @@ where
+            // 4. Pipe that stream as the body to the request to the history server
+            reqwest::Body::wrap_stream(stream)
+        };
+        #[cfg(target_arch = "wasm32")]
+        let body = {
+            use futures::AsyncReadExt;
+            // Make exporter mutable
+            let mut exporter = exporter;
+
+            // Wasm does not support stream uploads. So we'll just consume the stream into a vec.
+            let mut buffer = Vec::new();
+            exporter.read_to_end(&mut buffer).await?;
+            buffer
+        };
+
+        // 5. Make the request
+        let url = format!("{device_sync_server_url}/upload");
+        tracing::info!("Uploading sync payload to history server...");
+        let response = reqwest::Client::new().post(url).body(body).send().await?;
+        tracing::info!("Done uploading sync payload to history server.");
 
-        if !response.status().is_success() {
+        if let Err(err) = response.error_for_status_ref() {
             tracing::error!(
                 inbox_id = self.inbox_id(),
                 installation_id = hex::encode(self.installation_public_key()),
-                "Failed to upload file. Status code: {} Response: {response:?}",
-                response.status()
+                "Failed to upload file. Status code: {:?}",
+                err.status()
             );
-            response.error_for_status()?;
-            // checked for error, the above line bubbled up
-            unreachable!();
+            return Err(DeviceSyncError::Reqwest(err));
         }
 
-        let url = format!("{url}/files/{}", response.text().await?);
+        // Build a sync reply message that the new installation will consume
+        let reply = DeviceSyncReplyProto {
+            key,
+            request_id,
+            url: format!("{device_sync_server_url}/files/{}", response.text().await?),
+            metadata: Some(metadata),
 
-        let sync_reply = DeviceSyncReplyProto {
-            encryption_key: Some(enc_key.into()),
-            request_id: request_id.to_string(),
-            url,
-            timestamp_ns: now_ns() as u64,
-            kind: kind as i32,
+            // Deprecated fields
+            ..Default::default()
         };
 
-        Ok(sync_reply)
+        // Check acknowledgement one more time before responding to try to avoid double-responses
+        // from two or more old installations.
+        match acknowledge().await {
+            Err(DeviceSyncError::AlreadyAcknowledged) => {
+                return Ok(());
+            }
+            result => result?,
+        }
+
+        // Send the message out over the network
+        let content = DeviceSyncContent::Payload(reply);
+        self.send_device_sync_message(&provider, content, retry)
+            .await?;
+
+        handle.increment_metric(SyncMetric::PayloadSent);
+
+        Ok(())
     }
 
-    async fn insert_encrypted_syncables(
+    fn did_this_installation_ask_for_this_reply(
```

Comment:
IMO 'this' is a bit ambigious at first glance, i would personally prefer something like `is_reply_requested_by_installation` or simply `is_reply_requested` since a single installation should be implied by the nature of being part of `Client`. that's just a preference though so minor nit/feel free to ignore

Reply:
For sure. Makes much more sense to start with `is_` when it returns a bool.